File: flutter-boilerplate/test/lib/tier_features_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_boilerplate/api/server/billing/plan_prices.dart';
import 'package:flutter_boilerplate/l10n/app_localizations.dart';
import 'package:flutter_boilerplate/lib/tier_features.dart';
import 'package:flutter_test/flutter_test.dart';

// CROSS-031: the feature list comes from the backend, the words from ARB.
void main() {
  late AppLocalizations t;

  setUpAll(() async {
    t = await AppLocalizations.delegate.load(const Locale('en'));
  });

  test('translates known keys and interpolates values', () {
    expect(
      tierFeatureLabel(t, const TierFeature(key: 'basicAccess')),
      'Basic access',
    );
    expect(
      tierFeatureLabel(
        t,
        const TierFeature(key: 'everythingIn', value: 'BASIC'),
      ),
      'Everything in Basic',
    );
    expect(
      tierFeatureLabel(t, const TierFeature(key: 'callMinutes', value: '45')),
      'Calls up to 45 min',
    );
  });

  test('never drops an unknown key from a newer backend', () {
    expect(tierFeatureLabel(t, const TierFeature(key: 'teleport')), 'teleport');
    expect(
      tierFeatureLabel(t, const TierFeature(key: 'teleport', value: '3')),
      'teleport: 3',
    );
  });

  test('prefers the live backend list and falls back per tier', () {
    final prices = [
      const PlanPrice(
        tier: 'FREE',
        priceCents: 0,
        currency: 'USD',
        features: [
          TierFeature(key: 'basicAccess'),
          TierFeature(key: 'callMinutes', value: '10'),
        ],
      ),
      const PlanPrice(tier: 'BASIC', priceCents: 999, currency: 'USD'),
    ];
    expect(
      featuresForTier(t, 'free', prices),
      ['Basic access', 'Calls up to 10 min'],
    );
    // BASIC came back with no list → placeholder keys, still translated.
    expect(featuresForTier(t, 'basic', prices).first, 'Everything in Free');
    expect(featuresForTier(t, 'premium', null), contains('Crown badge'));
  });
}

File: next-js-boilerplate/src/lib/checkout/tier-features.test.ts
import { describe, it, expect } from "vitest";
import { featuresFromPlanPrices, tierFeatureLabel } from "./tier-features";

const labels = {
  basicAccess: "Basic access",
  everythingIn: "Everything in {tier}",
  callMinutes: "Calls up to {value} min",
  storageMultiplier: "{value}× storage",
};
const fallback = {
  FREE: ["fb-free"],
  BASIC: ["fb-basic"],
  MEDIUM: ["fb-medium"],
  PREMIUM: ["fb-premium"],
};

// CROSS-031: the list comes from the backend, the wording from i18n.
describe("tierFeatureLabel", () => {
  it("interpolates values and resolves everythingIn to the tier label", () => {
    expect(tierFeatureLabel({ key: "basicAccess" }, labels)).toBe(
      "Basic access",
    );
    expect(tierFeatureLabel({ key: "callMinutes", value: "45" }, labels)).toBe(
      "Calls up to 45 min",
    );
    expect(
      tierFeatureLabel({ key: "everythingIn", value: "BASIC" }, labels),
    ).toBe("Everything in Basic");
  });

  it("never drops an unknown key from a newer backend", () => {
    expect(tierFeatureLabel({ key: "teleport" }, labels)).toBe("teleport");
    expect(tierFeatureLabel({ key: "teleport", value: "3" }, labels)).toBe(
      "teleport: 3",
    );
  });
});

describe("featuresFromPlanPrices", () => {
  it("uses the backend list per tier and the fallback only where none arrived", () => {
    const result = featuresFromPlanPrices(
      [
        {
          tier: "FREE",
          priceCents: 0,
          currency: "USD",
          features: [
            { key: "basicAccess" },
            { key: "callMinutes", value: "10" },
          ],
        },
        { tier: "BASIC", priceCents: 999, currency: "USD", features: [] },
      ],
      labels,
      fallback,
    );
    expect(result.FREE).toEqual(["Basic access", "Calls up to 10 min"]);
    expect(result.BASIC).toEqual(["fb-basic"]);
    expect(result.PREMIUM).toEqual(["fb-premium"]);
  });

  it("returns the fallback verbatim before the query resolves", () => {
    expect(featuresFromPlanPrices(undefined, labels, fallback)).toEqual(
      fallback,
    );
  });
});
